### PR TITLE
[SPARK-13366] Support Cartesian join for Datasets

### PR DESCRIPTION
**Pull request by xguo27**

### CODE REVIEW

**Comment by marmbrus** [comment]
ok to test

File: sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala [hunk 1]
```diff
@@ -680,6 +681,14 @@ class Dataset[T] private[sql](
     joinWith(other, condition, "inner")
   }
 
+  /**
+   * Joins this [[Dataset]] returning a [[Tuple2]] for each pair using cartesian join
+   * Note: cartesian joins are very expensive without a filter that can be pushed down.
+   *
+   * @since 2.0.0
+   */
+  def joinWith[U](other: Dataset[U]): Dataset[(T, U)] = joinWith(other, lit(true), "inner")
```
Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: You don't actually need to use a true literal here, just construct a join without a condition.

Comment (xguo27) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: Thanks for your feedback @marmbrus . The only join API in Dataset I can find is:

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala#L644

which expects a Column. Do you mean to add some other method like the one in Dataframe:

https://github.com/apache/spark/blob/master/sql/core/src/main/scala/org/apache/spark/sql/DataFrame.scala#L383-L385

If so, I'm wondering whether we need to refactor out the code that handles encoder?

Comment (marmbrus) [inline sql/core/src/main/scala/org/apache/spark/sql/Dataset.scala]: I guess I would just make a protected version of `joinWith` above where the condition is an `Option` and have all the other methods delegate to that.

**Comment by xguo27** [comment]
Thanks @marmbrus ! I have updated the change following your suggestion.

**Comment by marmbrus** [comment]
@liancheng is this going to be done by your unification?  Should we merge or hold off?

**Comment by rxin** [comment]
Thanks for the pull request. I'm going through a list of pull requests to cut them down since the sheer number is breaking some of the tooling we have. Due to lack of activity on this pull request, I'm going to push a commit to close it. Feel free to reopen it or create a new one. We can also continue the discussion on the JIRA ticket.

**Pull request closed without merging**